src/Shared/Dictionary/SmallCapacityDictionary.cs

```diff
@@ -0,0 +1,761 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNetCore.Internal.Dictionary
+{
+    /// <summary>
+    /// An <see cref="IDictionary{String, Object}"/> type for route values.
+    /// </summary>
+    internal class SmallCapacityDictionary<TKey, TValue> : IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue> where TKey : notnull
+    {
+        // Threshold for size of array to use.
+        private static readonly int DefaultArrayThreshold = 4;
```

Comment:
10?

Reply:
I picked 4 based on the RouteValueDictionary and your comment about how going from 4 to 5 locals in state machines, it goes from fields to a list.

Reply:
Does eventually get slower:

```
|               Method |     Mean |     Error |    StdDev |   Median |      Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------- |---------:|----------:|----------:|---------:|----------:|------:|------:|------:|----------:|
|   OneValue_SmallDict | 1.212 us | 0.2310 us | 0.6702 us | 1.000 us | 824,829.9 |     - |     - |     - |      48 B |
|        OneValue_Dict | 2.113 us | 0.3505 us | 1.0169 us | 2.000 us | 473,170.7 |     - |     - |     - |      48 B |
| FourValues_SmallDict | 1.709 us | 0.2412 us | 0.6959 us | 1.400 us | 585,009.1 |     - |     - |     - |      48 B |
|      FourValues_Dict | 1.645 us | 0.2086 us | 0.5919 us | 1.700 us | 607,843.1 |     - |     - |     - |      48 B |
|  TenValues_SmallDict | 3.130 us | 0.2537 us | 0.7318 us | 2.900 us | 319,467.6 |     - |     - |     - |      48 B |
|       TenValues_Dict | 1.923 us | 0.2526 us | 0.7247 us | 1.900 us | 519,978.1 |     - |     - |     - |      48 B |
|            SmallDict | 1.745 us | 0.2066 us | 0.5478 us | 1.500 us | 573,025.9 |     - |     - |     - |     136 B |
|                 Dict | 2.524 us | 0.2664 us | 0.7065 us | 2.350 us | 396,135.3 |     - |     - |     - |     264 B |
|         SmallDictTen | 1.993 us | 0.2729 us | 0.7517 us | 1.750 us | 501,710.4 |     - |     - |     - |     280 B |
|              DictTen | 2.738 us | 0.4177 us | 1.1435 us | 2.200 us | 365,239.3 |     - |     - |     - |     488 B |
```